refactor: iframe plugin support

edoardo opened the pull request:
**Fixes [DHIS2-14216](https://dhis2.atlassian.net/browse/DHIS2-14216)**

### Key features

1. Extend the dashboard item search to include the `event visualizations`.

2. Implement the new secure iframe plugin solution needed for LL, DV and Map plugins and the other ones that will follow.
These dashboard items are rendered as `iframe` tags pointing to the location of the plugin.
Normally the URL used is the one specified in `pluginLaunchUrl` returned by the `apps` API.
There is a backdoor which is enabled only in dev mode where it's possible to override those locations via `localStorage` for development/testing purposes (ie. pointing a plugin to a port on localhost where the plugin runs).

Here's an example of override for both DV and LL plugins:
<img width="1083" alt="Screenshot 2022-11-09 at 16 17 47" src="https://user-images.githubusercontent.com/150978/200868964-4e11eb96-ea5d-4fc9-b925-4f4384bb88b0.png">


[DHIS2-14216]: https://dhis2.atlassian.net/browse/DHIS2-14216?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

edoardo committed:
chore: convert to functional component

edoardo committed:
feat: enable EVENT_VISUALIZATION dashboard item types

edoardo committed:
feat: allow iframe plugin overrides for dev/test

edoardo committed:
feat: render LL plugin in iframe

edoardo committed:
chore: add dependency for post-robot

edoardo committed:
chore: update pot file

edoardo committed:
fix: lookup in installed apps for the plugin URL to use in the iframe

edoardo committed:
chore: remove TODO comment

edoardo committed:
chore: update pot file

edoardo committed:
chore: add handy deduplicate yarn command

edoardo committed:
fix: fix ui dependency

edoardo committed:
chore: run prettier

jenniferarnesen commented on src/components/Item/VisualizationItem/ItemContextMenu/ViewAsMenuItems.js:
```diff
@@ -47,19 +48,23 @@ const ViewAsMenuItems = ({
-                    icon={<IconTable16 />}
-                />
-            )}
+            {activeType !== REPORT_TABLE &&
```
With the 3rd check, maybe better to switch to 
`![REPORT_TABLE, EVENT_REPORT, EVENT_VISUALIZATIOn].includes(activeType)`

What do you think?

edoardo replied:
Changed.

jenniferarnesen commented on src/components/Item/VisualizationItem/Visualization/Visualization.js:
```diff
@@ -21,100 +22,113 @@ import memoizeOne from '../memoizeOne.js'
-            getFilteredVisualization
-        )
-        this.memoizedGetVisualizationConfig = memoizeOne(getVisualizationConfig)
+    const getFilterVersion = memoizeOne(() => uniqueId())
```
Can you confirm that the memoization is still working correctly? I seem to remember leaving this component as a class because of issues with memoization. But I could be mis-remembering.

jenniferarnesen replied:
Also, if switching to functional component, I wonder if `useMemo` would work.

edoardo replied:
I didn't test the filters yet, since we don't know if we need to disable them for LL.
But good pointing this out, I wouldn't have known.

jenniferarnesen replied:
Actually, I think one of the issues could have been the amount of rendering. If you edit a dashboard and resize or move an item, is there unnecessary re-rendering/painting of any of the visualizations going on?

jenniferarnesen replied:
Also look for unnecessary re-rendering in view mode when changing View As, opening the interpretations, etc.

jenniferarnesen commented on src/components/Item/VisualizationItem/Visualization/IframePlugin.js:
```diff
@@ -0,0 +1,106 @@
+        const listener = postRobot.on(
+            'getProps',
+            // listen for messages coming only from the iframe rendered by this component
+            { window: iframeRef.current.contentWindow },
```
Any reason to be concerned that `iframRef` wouldn't exist yet?

edoardo replied:
Once `Suspense` is removed I don't think so.

edoardo replied:
Removed `Suspense`

jenniferarnesen commented on src/components/Item/VisualizationItem/Visualization/IframePlugin.js:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    return (
+        <Suspense fallback={<div />}>
```
Will there be a loading spinner like the other visualizations?

edoardo replied:
I think the plugin itself might already have a spinner.
But I don't think we need `Suspense` here, the iframe tag can be rendered right away, the problem is that until the plugin is mounted and got all the props it needs, nothing will be rendered inside the iframe.
This might not be a problem as it does not even depend on network, but there might be other reasons why the iframe content takes time before rendering anything.

edoardo replied:
Removed `Suspense`

edoardo replied:
Clarification:
the new `PluginWrapper` components in the various apps show a loading spinner while the communication between dashboard app and iframe plugin content happen.

edoardo committed:
refactor: rewrite condition for better readability

edoardo committed:
fix: remove Suspense which might interfere with the iframe

edoardo committed:
fix: ensure the selector content fits in the container

This is a ui issue: https://github.com/dhis2/ui/pull/1152.

edoardo committed:
refactor: use Layer + Popper to avoid bottom double border

edoardo committed:
fix: remove redundant overflow rule, fixed in ui

This reverts commit a36ab37ae0d5aed0495d0fc365fe4da5236594d4.

edoardo committed:
chore: deduplicate yarn.lock

edoardo committed:
refactor: preserve height/width if provided, update props passed

edoardo committed:
refactor: use iframe for DV plugin

edoardo committed:
chore: remove unused import

KaiVandivier committed:
feat: add recording props to plugin (#2125)

* fix(pwa): make cache filters more specifc

* feat(plugins): add `record` prop to iframe plugins when caching dashboard

* feat: remove cached plugin data with dashboard

edoardo committed:
chore: run Prettier

edoardo committed:
test: fix failing tests

edoardo committed:
chore: remove unused code

edoardo committed:
chore: run Prettier

edoardo committed:
fix: pass correct prop for display property

Also do not pass the whole userSettings object, not used in the plugins
where the only thing really needed is the displayProperty for the
analytics request.

edoardo committed:
chore: deduplicate yarn.lock

edoardo committed:
fix: add more paths to omit from cache

Replaced deprecated config key with new one.

edoardo committed:
fix: show error when iframe src is not available

Also avoid adding postRobot handlers when the iframeRef is not
available, this removes some extra errors.

edoardo committed:
fix: do not pass filtered AO to LL plugin

Filters don't work in LL without modifications.
For now just disable filtering for LL items.

edoardo committed:
fix: set error when pluginLaunchUrl is not available

The app could be installed and the entry in apps api returned, but the
plugin launch URL can be missing, in that case the plugin cannot be
loaded in the iframe.

edoardo committed:
refactor: use specific reducer selector

edoardo committed:
fix: remove View As in context menu for LL items

edoardo committed:
fix: clear selected from Redux when editing a new dashboard

edoardo committed:
feat: add tags on item header, used for filters in LL

For LL items we don't apply filters.
Show a tag in the item header to indicate this to the user.

edoardo committed:
feat: add overlay on LL items to inform about not applied filters

edoardo committed:
refactor: move the overlay handling to the parent component

In this way both the tag in the item header and the overlay in the item
content can be controlled at the same time allowing for toggling the tag
depending on whether the overlay is shown or not.

edoardo committed:
test: update snapshots

edoardo committed:
fix: remove tag icon and fix styles

Approved by Joe.

edoardo committed:
chore: update pot file

edoardo committed:
fix: remove context menu when plugin is not installed

* fetch schema for eventVisualizations, required by the interpretations
  component

edoardo committed:
fix: rename Event visualizations to Line lists

This appears as section title in the dashboard item search.

edoardo committed:
feat: add custom message for when the plugin is not installed

More friendly than the generic error message.
In this case there isn't any problem with the dashboard item, but the
plugin for rendering the item is not installed.
We inform the user about it and suggest to install the missing app from
the app hub.

edoardo committed:
refactor: propagate d2 needed to check installed apps

edoardo committed:
feat: use iframe for Map items

edoardo committed:
chore: update d2-ui-interpretations component

edoardo committed:
fix: update iframe src and props when using View As

edoardo committed:
chore: fix linting issue

edoardo committed:
fix: fix re-render of visualizations when interpretations panel toggles (#2199)

edoardo committed:
fix: add app id for DV when fetched from installed apps

edoardo committed:
fix: detect plugin availability via plugin launch URL

edoardo committed:
fix: add fallback to bundled plugin path for core apps

This addresses the case of core apps plugins which are always served
from the bundled app's path, instead of from api/apps.

For core (bundled) apps:
* first we look at the api/apps, if there is an entry there for the
  app it means it has been manually installed from the app hub
  overriding the bundled version.
  The plugin URL in this case is not in the format api/apps but
  dhis-web-*.
* if no entry is returned from api/apps we use the "hardcoded" path (as
  before) thus using the bundled version of the plugin.

This should work for situations where a new version of a core app is
installed or uninstalled from the app hub.

edoardo committed:
chore: remove accidentally committed temp files

edoardo committed:
fix: use base url without api version

This should solve the issue of bundled apps plugin not working in
iframes.

edoardo committed:
fix: deduplicate yarn.lock

edoardo committed:
fix: attempt at fixing the sizing issue in DV iframe content

jenniferarnesen committed:
feat: map plugin (#2229)

jenniferarnesen commented on src/pages/edit/NewDashboard.js:
```diff
@@ -2,9 +2,10 @@ import i18n from '@dhis2/d2-i18n'
 import PropTypes from 'prop-types'
 import React, { useState, useEffect } from 'react'
-import { connect } from 'react-redux'
+import { connect, useDispatch } from 'react-redux'
```
What about switching connect with useSelector, for consistency?
